Alerting section audit - Part 1 (#5407)
DataDog/documentation: content/en/monitors/faq/anomaly-monitor.md

---
title: Anomaly Monitor FAQ
kind: faq
---

### Should I use anomaly detection for everything?

> No. Anomaly detection is designed to assist with visualizing and monitoring metrics that have predictable patterns. For example, `my_site.page_views{*}` might be driven by user traffic and thus vary predictably by time of day and day of week. If your metric does not have a repeated or predictable pattern, then a simple chart overlay or threshold alert might be better.

Reviewer: ```suggestion
No. Anomaly detection is designed to assist with visualizing and monitoring metrics that have predictable patterns. For example, you wouldn't want to use Anomaly detection if `my_site.page_views{*}` is driven by user traffic that varies predictably by time of day and day of week. If your metric has a repeated or predictable pattern, then a chart overlay or threshold alert might be better.
```

this last part seems like it should be if it does have a predictable pattern, but I am not sure. This example is oddly phrased, so I am just trying to make it a bit more clear with my suggestion.

Author: Agreed. `my_site.page_views` is a metric that works well with anomaly detection. I will reword this.

**Note**: Anomaly detection requires historical data to make good predictions. If you have only been collecting a metric for a few hours or days, anomaly detection won't be useful.

### Why can't I use anomaly detection over groups in the dashboard?

Adding many separate timeseries in a single graph can lead to [spaghettification][1], and the problem gets worse when the anomaly detection visualization is added:

{{< img src="monitors/monitor_types/anomaly/spaghetti.png" alt="spaghetti" responsive="true" style="width:80%;">}}

However, it is possible to add multiple series to a single graph one at a time. The gray band only shows up on mouseover:

{{< img src="monitors/monitor_types/anomaly/anomaly_multilines.png" alt="anomaly multi lines" responsive="true" style="width:80%;" >}}

### Do past anomalies affect the current predictions?

> All the algorithms except `basic` use extensive amounts of historical data so they are robust to most anomalies. In the first graph, the gray band stays around 400K even after the metric has dropped to 0.

Reviewer: ```suggestion
All the algorithms except `basic` use extensive amounts of historical data, so they are robust to most anomalies. In the first graph, the gray band stays around 400K even after the metric has dropped to 0.
```

{{< img src="monitors/monitor_types/anomaly/anomalous_history.png" alt="anomalous_history" responsive="true" style="width:80%;">}}

The second graph shows the same metric a day later. Even though it uses the previous day in the calculation of the band, it is unaffected by the anomaly that occurred previously.

{{< img src="monitors/monitor_types/anomaly/no_effect.png" alt="no effect" responsive="true" style="width:80%;">}}

### Why does an anomaly "disappear" when I zoom in?

At different zoom levels, the same query can result in timeseries with very different characteristics. When looking at longer time periods, each point represents the aggregate of many more-granular points. Therefore, each of these aggregate points may hide noise observed in the more granular points. For example, charts that show one week often appear smoother (less noisy) than charts that show just 10 minutes.

> The width of the gray band is partly based on the noisiness of the timeseries in the plot. The band must be wide enough that ordinary noise is mostly inside the band and doesn't appear as anomalous. Unfortunately, when the band is wide enough to include ordinary noise, it might also be wide enough to hide some anomalies, especially when viewing short time windows.

Reviewer: ```suggestion
Setting the width of the gray band correctly is key to accurate anomaly monitoring. The band must be wide enough that ordinary noise is mostly inside the band and doesn't appear as anomalous. Unfortunately, when the band is wide enough to include ordinary noise, it might also be wide enough to hide some anomalies, especially when viewing short time windows.
```

Author: The user cannot set the width of the gray band. I will reword this a little.

In general, if an anomaly disappears when you zoom in, this doesn't mean that it's not an anomaly. It means, while the individual points in the zoomed-in view are not anomalous in isolation, many slightly unusual points occurring together is anomalous.

### Why do I get a query parsing error when trying to combine some functions with anomaly detection?

Not all functions may be nested inside of calls to the `anomalies()` function. In particular, you may not include any of the following functions in an anomaly detection monitor or dashboard query: `cumsum()`, `integral()`, `outliers()`, `piecewise_constant()`, `robust_trend()`, or `trend_line()`.

Anomaly detection uses historical data to establish a baseline of normal behavior for a series. The functions above are sensitive to the placement of the query window. The value of the series at a single timestamp can change significantly based upon where it falls within the query window. This sensitivity prevents anomaly detection from determining a consistent baseline for the series.

### What happened to the adaptive algorithm?

> Previously, Datadog used an algorithm called `adaptive`, which would try to figure out a metric's inherent seasonality and adjust the predictions accordingly. Now, the seasonality is automatically detected when setting up a monitor. There is less need for this algorithm, which was slower and required more data than the other algorithms. Existing monitors that use the `adaptive` algorithm are untouched and continue to work.

Reviewer: ```suggestion
Datadog has evolved our algorithms to no longer need to use the adaptive algorithm. Existing monitors that use the `adaptive` algorithm are untouched and continue to work.
```
the key information here is that existing monitors still work. otherwise we should stay away from explaining our dev decisions

Author: Sounds good. Updating to change the wording slightly.

### What is the count_default_zero argument?

> Previously, Datadog treated count metrics as gauges, and thus interpolated between reported points. This led to some very odd-looking metrics for sparsely reported counts. Anomalies are no longer interpolating between counts, but for legacy monitors, the old behavior is preserved using the `count_default_zero` argument.

Reviewer: same thing here

Create **A**: an anomaly monitor to alert on values above the bounds; and **B**: a separate [metric monitor][2] with a threshold alert to trigger on values greater than X. Then create a [composite monitor][3] on **A && B**.

### Why am I prevented from saving a monitor with the message "alert and alert recovery criteria are such that the monitor can be simultaneously in alert and alert recovery states"?

Setting different windows for the alert and alert recovery periods might lead to an ambiguous state. The alert and alert recovery window sizes should be set such that both cannot be satisfied at the same time. For example, setting an alert threshold at 50% for a 2-hour window (i.e., 1 hour has to be anomalous to trigger the alert) and the [recovery threshold][4] at 50% for a 10-minute window (i.e., 5 minutes have to be non-anomalous to recover) might result in triggering the alert and the alert recovery states simultaneously. If the last 5 minutes are not anomalous but the 1 hour before _was_ anomalous, both the alert and the alert recovery are triggered.

### How does daylight savings affect anomaly detection monitors?

> Datadog monitors use UTC time and by default are agnostic to local time zones (e.g. EST, PST, CST). User activity is shifted relative to UTC time because activity typically remains the same for the user's local time. This could be detected as unexpected anomaly. 

Reviewer: ```suggestion
Datadog monitors use UTC time and by default are agnostic to local time zones. User activity is shifted relative to UTC time because activity typically remains the same for the user's local time. This could be detected as an unexpected anomaly. 
```
